app/Http/Controllers/App/UsersController.php

```diff
@@ -36,58 +48,176 @@ public function add(Request $request)
         $user->save();
+        DB::commit();
 
         return self::json($user->toArrayCamelize(), 200);
     }
 
     public function emailActivate(Request $request)
     {
-        $this->validateRequest($request, 'user', User::$rules_email_activate);
-
-        $user = User::where(
-            'email_confirm_token',
-            $request->input('user.email_confirm_token')
-        )->whereNull('email_confirmed_at')->first();
+        Validator::make($request->all(), ['user.email_confirm_token' => 'required'])->validate();
 
+        DB::beginTransaction();
+        if (false === $token = Token::check($request->input('user.email_confirm_token'))) {
+            return self::json([], 403);
+        }
+        $user = User::find($token['user_id']);
         if (empty($user)) {
-            return self::json([], 401);
+            return self::json([], 403);
         }
-
         $user->email_confirmed_at = date('Y-m-d H:i:s');
         $user->save();
+        DB::commit();
+
+        return self::json($user->toArrayCamelize(), 200);
+    }
+
+    public function emailActivateResend(Request $request)
+    {
+        Validator::make($request->all(), ['uri' => 'required'])->validate();
+        $user = Auth::user();
+        DB::beginTransaction();
+        if (!Token::canGenerate($user->id, 'email-activate', $this->email_activation_resend_limit)) {
+            return self::json([], 400, ['message' => 'You can request 1 email activation every 15 minutes. Please wait.']);
+        }
+        Mail::to($user)->queue(new UserEmailActivate(
+            Token::generate('email-activate', $this->email_activation_token_time, $user->id),
+            $request->input('uri')
+        ));
+        DB::commit();
+
+        return self::json([], 204);
+    }
+
+    public function emailChangeStep1(Request $request)
+    {
+        Validator::make($request->all(), ['email' => 'required|email', 'URIstep1' => 'required', 'URIstep2' => 'required'])->validate();
+        if (User::withTrashed()->where('email', $request->input('email'))->count()) {
+            return self::json([], 422, ['email' => 'This email already exists in our database']);
+        }
+
+        $user = Auth::user();
+        DB::beginTransaction();
+        Mail::to($user)->queue(new UserEmailChangeConfirm1Old(
+            Token::generate('email-change-step1', $this->email_change_token_time, $user->id, $request->all()),
+            $request->input('URIstep1')
+        ));
+        DB::commit();
+
+        return self::json([], 204);
+    }
+
+    public function emailChangeStep2($token)
+    {
+        DB::beginTransaction();
+        if (false === $token = Token::check($token)) {
+            DB::commit();
+
+            return self::json([], 403, ['message' => 'Invalid token or outdated']);
+        }
+        $user = User::findOrFail($token['user_id']);
+        if (User::withTrashed()->where('email', $token['payload']['email'])->count()) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "payload" 5 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=php%3AS1192)